addons/purchase_stock/models/purchase.py

```diff
@@ -252,14 +259,39 @@ def _create_picking(self):
         for order in self.filtered(lambda po: po.state in ('purchase', 'done')):
             if any(product.type in ['product', 'consu'] for product in order.order_line.product_id):
                 order = order.with_company(order.company_id)
-                pickings = order.picking_ids.filtered(lambda x: x.state not in ('done', 'cancel'))
-                if not pickings:
-                    res = order._prepare_picking()
-                    picking = StockPicking.with_user(SUPERUSER_ID).create(res)
-                    pickings = picking
-                else:
-                    picking = pickings[0]
-                moves = order.order_line._create_stock_moves(picking)
+                pickings = self.env["stock.picking"]
+                processed_pickings = self.env["stock.picking"]
+                receipt_lines = order.order_line.filtered(
+                    lambda l: l.product_qty >= 0 and l.product_id.type in ['product', 'consu']
+                )
+                return_lines = order.order_line.filtered(
+                    lambda l: l.product_qty < 0 and l.product_id.type in ['product', 'consu']
+                )
+                moves = self.env['stock.move']
+                if receipt_lines:
+                    receipt_pickings = order.picking_ids.filtered(
+                        lambda p: p.state not in ('done', 'cancel') and p.picking_type_id.code == 'incoming'
+                    )
+                    if not receipt_pickings:
+                        picking_vals = order._prepare_picking()
+                        receipt_picking = StockPicking.with_user(SUPERUSER_ID).create(picking_vals)
+                    else:
+                        receipt_picking = receipt_pickings[0]
+                    pickings += receipt_pickings
+                    processed_pickings += receipt_picking
+                    moves += receipt_lines._create_stock_moves(receipt_picking)
+                if return_lines:
+                    return_pickings = order.picking_ids.filtered(
+                        lambda p: p.state not in ('done', 'cancel') and p.picking_type_id.code != 'incoming'
+                    )
+                    if not return_pickings:
+                        return_picking_vals = order._prepare_picking(is_return=True)
+                        return_picking = StockPicking.with_user(SUPERUSER_ID).create(return_picking_vals)
+                    else:
+                        return_picking = return_pickings[0]
+                    processed_pickings += return_picking
+                    pickings += return_pickings
+                    moves += return_lines._create_stock_moves(return_picking, is_return=True)
```

Comment:
The lines are repetitive here. We should create a common method to be called from each condition.